src/blob_store/store_impl.rs

```diff
@@ -0,0 +1,372 @@
+use crate::packet::{Blob, BlobError, BLOB_HEADER_SIZE};
+use crate::result::Error as SErr;
+
+use byteorder::{BigEndian, ByteOrder, WriteBytesExt};
+
+use std::borrow::Borrow;
+use std::collections::HashMap;
+use std::fs::{self, File, OpenOptions};
+use std::io::{prelude::*, BufReader, BufWriter};
+use std::path::{Path, PathBuf};
+
+use super::*;
+
+const DATA_FILE_NAME: &str = "data";
+pub(super) const META_FILE_NAME: &str = "meta";
+const INDEX_FILE_NAME: &str = "index";
+pub(super) const ERASURE_FILE_NAME: &str = "erasure";
+pub(super) const ERASURE_INDEX_FILE_NAME: &str = "erasure_index";
+
+//const DATA_FILE_BUF_SIZE: usize = 2 * 1024 * 1024;
+const DATA_FILE_BUF_SIZE: usize = 64 * 1024;
+
+pub(super) const INDEX_RECORD_SIZE: u64 = 3 * 8;
+
+impl Store {
+    pub(super) fn mk_slot_path(&self, slot_height: u64) -> PathBuf {
+        let splat = slot_height.to_be_bytes();
+        let mut path = self.root.join(format!("{:#04x}", splat[0]));
+
+        for byte in &splat[1..] {
+            path = path.join(format!("{:02x}", byte));
+        }
+
+        path
+    }
+
+    pub(super) fn mk_data_path(&self, slot_height: u64) -> PathBuf {
+        self.mk_slot_path(slot_height).join(DATA_FILE_NAME)
+    }
+
+    pub(super) fn mk_index_path(&self, slot_height: u64) -> PathBuf {
+        self.mk_slot_path(slot_height).join(INDEX_FILE_NAME)
+    }
+
+    pub(super) fn mk_erasure_path(&self, slot_height: u64) -> PathBuf {
+        self.mk_slot_path(slot_height).join(ERASURE_FILE_NAME)
+    }
+
+    pub(super) fn mk_erasure_index_path(&self, slot: u64) -> PathBuf {
+        self.mk_slot_path(slot).join(ERASURE_INDEX_FILE_NAME)
+    }
+
+    // TODO: possibly optimize by checking metadata and immediately quiting based on too big indices
+    pub(super) fn index_data(
+        &self,
+        slot_height: u64,
+        blob_index: u64,
+    ) -> Result<(PathBuf, BlobIndex)> {
+        let slot_path = self.mk_slot_path(slot_height);
+        if !slot_path.exists() {
+            return Err(StoreError::NoSuchSlot(slot_height));
+        }
+
+        let (data_path, index_path) = (
+            slot_path.join(DATA_FILE_NAME),
+            slot_path.join(INDEX_FILE_NAME),
+        );
+
+        let mut index_file = BufReader::new(File::open(&index_path)?);
+
+        let mut buf = [0u8; INDEX_RECORD_SIZE as usize];
+        while let Ok(_) = index_file.read_exact(&mut buf) {
+            let index = BigEndian::read_u64(&buf[0..8]);
+            if index == blob_index {
```

Comment:
this code takes N/2 to find the index record, it could be constant time

Reply:
Wouldn't I have to ensure that indexes were written in-order to make that possible?

Reply:
no, you can seek() to the spot you need to write() at in the index file.  the file will grow automatically, and will be filled with zeros (which you can call "invalid" offsets)

Reply:
Got it. I'm already doing this in another place, will make this change tomorrow